RuntimeException: javax.crypto.BadPaddingException

We have this exception from time to time. I suppose Stormpath should be handling this as a failed authentication instead of throwing it as a runtime exception?

stacktrace for version: 1.0.RC7.4

```
java.lang.RuntimeException: javax.crypto.BadPaddingException: Given final block not properly padded
        at com.stormpath.sdk.impl.security.ApiKeySecretEncryptionService.decrypt(ApiKeySecretEncryptionService.java:124)
        at com.stormpath.sdk.impl.security.ApiKeySecretEncryptionService.decryptBase64String(ApiKeySecretEncryptionService.java:100)
        at com.stormpath.sdk.impl.ds.api.DecryptApiKeySecretFilter.clone(DecryptApiKeySecretFilter.java:128)
        at com.stormpath.sdk.impl.ds.api.DecryptApiKeySecretFilter.filter(DecryptApiKeySecretFilter.java:88)
        at com.stormpath.sdk.impl.ds.DefaultFilterChain.filter(DefaultFilterChain.java:52)
        at com.stormpath.sdk.impl.ds.EnlistmentFilter.filter(EnlistmentFilter.java:42)
        at com.stormpath.sdk.impl.ds.DefaultFilterChain.filter(DefaultFilterChain.java:52)
        at com.stormpath.sdk.impl.ds.DefaultDataStore.getResourceData(DefaultDataStore.java:286)
        at com.stormpath.sdk.impl.ds.DefaultDataStore.getResource(DefaultDataStore.java:201)
        at com.stormpath.sdk.impl.ds.DefaultDataStore.getResource(DefaultDataStore.java:197)
        at com.stormpath.sdk.impl.application.DefaultApplication.getApiKey(DefaultApplication.java:535)
        at com.stormpath.sdk.impl.authc.BasicApiAuthenticator.authenticate(BasicApiAuthenticator.java:55)
        at com.stormpath.sdk.impl.authc.BasicApiAuthenticator.authenticate(BasicApiAuthenticator.java:49)
        at com.stormpath.sdk.impl.oauth.authc.OauthAuthenticationRequestDispatcher.authenticate(OauthAuthenticationRequestDispatcher.java:54)
        at com.stormpath.sdk.impl.application.DefaultApplication.authenticateAccount(DefaultApplication.java:337)
        at com.stormpath.sdk.impl.authc.DefaultApiRequestAuthenticator.execute(DefaultApiRequestAuthenticator.java:82)
        at com.stormpath.sdk.impl.application.DefaultApplication.authenticateApiRequest(DefaultApplication.java:557)
        at ...
Caused by: javax.crypto.BadPaddingException: Given final block not properly padded
        at com.sun.crypto.provider.CipherCore.doFinal(CipherCore.java:966)
        at com.sun.crypto.provider.CipherCore.doFinal(CipherCore.java:824)
        at com.sun.crypto.provider.AESCipher.engineDoFinal(AESCipher.java:436)
        at javax.crypto.Cipher.doFinal(Cipher.java:2165)
        at com.stormpath.sdk.impl.security.ApiKeySecretEncryptionService.decrypt(ApiKeySecretEncryptionService.java:119)
        ... 59 more
```

--- a/impl/src/main/java/com/stormpath/sdk/impl/authc/BasicApiAuthenticator.java
+++ b/impl/src/main/java/com/stormpath/sdk/impl/authc/BasicApiAuthenticator.java
@@ -27,11 +27,15 @@ import com.stormpath.sdk.impl.api.DefaultApiKeyOptions;
 import com.stormpath.sdk.impl.ds.InternalDataStore;
 import com.stormpath.sdk.impl.error.ApiAuthenticationExceptionFactory;
 import com.stormpath.sdk.lang.Assert;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
 
 /**
  * @since 1.0.RC
  */
 public class BasicApiAuthenticator {
+
+    private static final Logger log = LoggerFactory.getLogger(BasicApiAuthenticator.class);
 
     private final InternalDataStore dataStore;
 
@@ -52,7 +56,24 @@ public class BasicApiAuthenticator {
     public ApiAuthenticationResult authenticate(Application application, String id, String secret) {
         Assert.notNull(application, "application  cannot be null.");
 
-        ApiKey apiKey = application.getApiKey(id, new DefaultApiKeyOptions().withAccount());
+        //Workaround for https://github.com/stormpath/stormpath-sdk-java/issues/440
+        ApiKey apiKey = null;
+        for (int i = 0; i < 3; i++) {
+            try {
+                apiKey = application.getApiKey(id, new DefaultApiKeyOptions().withAccount());
+                break;
+            } catch (Exception e) {
+                log.error("Couldn't get Application ApiKey for " + application.getHref(), e);
+                if (i == 2) {
+                    throw new RuntimeException(e);
+                }
+                try {
+                    Thread.sleep(500);
+                } catch (InterruptedException e1) {
+                    throw new RuntimeException(e1);
+                }
+            }
+        }
 
         if (apiKey == null || !apiKey.getSecret().equals(secret)) {
             throw ApiAuthenticationExceptionFactory.newApiAuthenticationException(IncorrectCredentialsException.class);

--- a/impl/src/test/groovy/com/stormpath/sdk/impl/authc/BasicApiAuthenticatorTest.groovy
+++ b/impl/src/test/groovy/com/stormpath/sdk/impl/authc/BasicApiAuthenticatorTest.groovy
@@ -1,0 +1,97 @@
+package com.stormpath.sdk.impl.authc
+
+import com.stormpath.sdk.account.Account
+import com.stormpath.sdk.account.AccountStatus
+import com.stormpath.sdk.api.ApiKey
+import com.stormpath.sdk.api.ApiKeyOptions
+import com.stormpath.sdk.api.ApiKeyStatus
+import com.stormpath.sdk.application.Application
+import com.stormpath.sdk.impl.ds.InternalDataStore
+import org.testng.annotations.BeforeMethod
+import org.testng.annotations.Test
+
+import javax.crypto.BadPaddingException
+
+import static org.easymock.EasyMock.anyObject
+import static org.easymock.EasyMock.createMock
+import static org.easymock.EasyMock.eq
+import static org.easymock.EasyMock.expect
+import static org.easymock.EasyMock.replay
+import static org.easymock.EasyMock.verify
+import static org.testng.Assert.assertNotNull
+
+/**
+ * @since 1.0.4
+ */
+class BasicApiAuthenticatorTest {
+
+    private static final String keyId = "keyId"
+    private static final String keySecret = "keySecret"
+
+    ApiKey mockApiKey
+    Application mockApplication
+    Account mockAccount
+    InternalDataStore mockDataStore
+
+    BasicApiAuthenticator authenticator
+
+    @BeforeMethod
+    void setup() {
+        mockApiKey = createMock(ApiKey)
+        mockApplication = createMock(Application)
+        mockAccount = createMock(Account)
+        mockDataStore = createMock(InternalDataStore)
+
+        authenticator = new BasicApiAuthenticator(mockDataStore)
+    }
+
+    @Test
+    void testAuthenticationNormalWorkflow() {
+        expect(mockApplication.getApiKey(eq(keyId), anyObject(ApiKeyOptions.class))).andReturn mockApiKey
+        expect(mockApiKey.getSecret()).andReturn keySecret
+        expect(mockApiKey.getStatus()).andReturn ApiKeyStatus.ENABLED
+        expect(mockApiKey.getAccount()).andReturn mockAccount
+        expect(mockAccount.getStatus()).andReturn AccountStatus.ENABLED
+
+        replay mockAccount, mockApiKey, mockApplication, mockDataStore
+
+        def result = authenticator.authenticate(mockApplication, keyId, keySecret)
+        assertNotNull(result)
+
+        verify mockAccount, mockApiKey, mockApplication, mockDataStore
+    }
+
+    @Test
+    void testAuthenticationOneFailure() {
+        expect(mockApplication.getApiKey(eq(keyId), anyObject(ApiKeyOptions.class))).andThrow new RuntimeException(new BadPaddingException())
+        expect(mockApplication.getHref()).andReturn "http://some_href"
+        expect(mockApplication.getApiKey(eq(keyId), anyObject(ApiKeyOptions.class))).andReturn mockApiKey
+        expect(mockApiKey.getSecret()).andReturn keySecret
+        expect(mockApiKey.getStatus()).andReturn ApiKeyStatus.ENABLED
+        expect(mockApiKey.getAccount()).andReturn mockAccount
+        expect(mockAccount.getStatus()).andReturn AccountStatus.ENABLED
+
+        replay mockAccount, mockApiKey, mockApplication, mockDataStore
+
+        def result = authenticator.authenticate(mockApplication, keyId, keySecret)
+        assertNotNull(result)
+
+        verify mockAccount, mockApiKey, mockApplication, mockDataStore
+    }
+
+    @Test(expectedExceptions = RuntimeException.class)
+    void testAuthenticationFailure() {
+        expect(mockApplication.getApiKey(eq(keyId), anyObject(ApiKeyOptions.class))).andThrow(new RuntimeException(new BadPaddingException())).times(3)
+        expect(mockApplication.getHref()).andReturn("http://some_href").times(3)
+        expect(mockApiKey.getSecret()).andReturn keySecret
+        expect(mockApiKey.getStatus()).andReturn ApiKeyStatus.ENABLED
+        expect(mockApiKey.getAccount()).andReturn mockAccount
+        expect(mockAccount.getStatus()).andReturn AccountStatus.ENABLED
+
+        replay mockAccount, mockApiKey, mockApplication, mockDataStore
+
+        authenticator.authenticate(mockApplication, keyId, keySecret)
+
+        verify mockAccount, mockApiKey, mockApplication, mockDataStore
+    }
+}

--- a/impl/src/test/groovy/com/stormpath/sdk/impl/authc/BasicApiAuthenticatorTest.groovy
+++ b/impl/src/test/groovy/com/stormpath/sdk/impl/authc/BasicApiAuthenticatorTest.groovy
@@ -13,7 +13,7 @@ import org.testng.annotations.Test
 import javax.crypto.BadPaddingException
 
 import static org.easymock.EasyMock.anyObject
-import static org.easymock.EasyMock.createMock
+import static org.easymock.EasyMock.createStrictMock
 import static org.easymock.EasyMock.eq
 import static org.easymock.EasyMock.expect
 import static org.easymock.EasyMock.replay
@@ -37,10 +37,10 @@ class BasicApiAuthenticatorTest {
 
     @BeforeMethod
     void setup() {
-        mockApiKey = createMock(ApiKey)
-        mockApplication = createMock(Application)
-        mockAccount = createMock(Account)
-        mockDataStore = createMock(InternalDataStore)
+        mockApiKey = createStrictMock(ApiKey)
+        mockApplication = createStrictMock(Application)
+        mockAccount = createStrictMock(Account)
+        mockDataStore = createStrictMock(InternalDataStore)
 
         authenticator = new BasicApiAuthenticator(mockDataStore)
     }
@@ -81,8 +81,12 @@ class BasicApiAuthenticatorTest {
 
     @Test(expectedExceptions = RuntimeException.class)
     void testAuthenticationFailure() {
-        expect(mockApplication.getApiKey(eq(keyId), anyObject(ApiKeyOptions.class))).andThrow(new RuntimeException(new BadPaddingException())).times(3)
-        expect(mockApplication.getHref()).andReturn("http://some_href").times(3)
+        expect(mockApplication.getApiKey(eq(keyId), anyObject(ApiKeyOptions.class))).andThrow(new RuntimeException(new BadPaddingException()))
+        expect(mockApplication.getHref()).andReturn("http://some_href")
+        expect(mockApplication.getApiKey(eq(keyId), anyObject(ApiKeyOptions.class))).andThrow(new RuntimeException(new BadPaddingException()))
+        expect(mockApplication.getHref()).andReturn("http://some_href")
+        expect(mockApplication.getApiKey(eq(keyId), anyObject(ApiKeyOptions.class))).andThrow(new RuntimeException(new BadPaddingException()))
+        expect(mockApplication.getHref()).andReturn("http://some_href")
         expect(mockApiKey.getSecret()).andReturn keySecret
         expect(mockApiKey.getStatus()).andReturn ApiKeyStatus.ENABLED
         expect(mockApiKey.getAccount()).andReturn mockAccount

